fix: create tnscope mnvs

mathiasbio opened the pull request:
## Description

<!-- Provide a brief overview of your PR and link any relevant user stories or issues. -->

This PR aims to add a post-processing step to TNscope (for standard TGA workflow) to merge SNVs and InDels by their PID to MNVs to be more accurately merged with the VarDict results.

See issue: https://github.com/Clinical-Genomics/BALSAMIC/issues/1525

The script was taken from Sentieon here: https://github.com/Sentieon/sentieon-scripts/blob/master/merge_mnp/merge_mnp.py

Script has been slightly modified to merge variants despite filters other filters than PASS, this to allow for merging of variants with triallelic-site, and any other future soft-filters we might add. Some refactoring was done as well from the original filter to increase readability. (though it's still very messy)

**The TNscope VCF is quality filtered before merging SNVs**

To avoid merging of low quality variant to MNVs, the VCF is quality filtered before merging.

**Merging SNVs with different filters set**

There is an issue about how to consolidate the filters when merging SNVs with different filters set. Such as: germline_risk, in_normal, triallelic_site, and PASS. This was solved with logic that can be exemplified in this table below:

| v1                        | v2              | v3            | filter_set                              |
|---------------------------|-----------------|---------------|-----------------------------------------|
| in_normal                 | PASS            | germline_risk | MNV_CONFLICTING_FILTERS                 |
| in_normal,germline_risk   | in_normal       |               | in_normal,germline_risk                 |
| in_normal                 | triallelic_site | germline_risk | MNV_CONFLICTING_FILTERS                 |
| in_normal,triallelic_site | triallelic_site |               | MNV_CONFLICTING_FILTERS                 |
| in_normal,triallelic_site | germline_risk   |               | in_normal,triallelic_site,germline_risk |
| in_normal,triallelic_site | PASS            |               | MNV_CONFLICTING_FILTERS                 |

In summary:
1. If constituent variants have mixed filters the filter is set to MNV_CONFLICTING_FILTERS
2. UNLESS, at least 1 of the "normal_filters", that is "in_normal" and "germline_risk" is set in ALL variants of the MNV, then the set of the filters are added instead
3. And finally of course, if the variants all have the same filter, such as PASS, then that filter is set.

On top of this, a few new INFO fields are added to preserve some information from the constituent variants, originally only the FILTER was added but this has been amended with:
- TUMOR_DP (ref|alt)
- NORMAL_DP (ref|alt)
- TUMOR_AF
- NORMAL_AF
- variant chrom_pos_ref_alt

With the values joined as comma-separated list for all variants. 

Regarding benchmarking the speed of the merge script:
- For most standard TGA cases that I checked the entire rule which includes multiple post-processing of TNscope takes 7-8 seconds
- For WES it took 22 seconds with a maximum memory peak of 50MB 


#### Added

- TNscope MNV merge script

#### Changed

- merge SNVs into MNVs in TNscope TGA
- change raw delivery SNV file for TGA to before any post-processing

## Documentation

<!-- Link all added or updated documents. -->

- [ ] N/A
- [x] Updated Balsamic documentation to reflect the changes as needed for this PR.
    - balsamic_filters.rst (updated with TNscope MNV merging)

## Tests

<!-- Describe in detail how you tested your changes to help reviewers validate the code. -->
<!-- Include screenshots or visual representations of your changes. -->

### Feature Tests

<!-- Include tests relevant to the changes in this PR. -->

- [ ] N/A
- [x] Test that variants are correctly merged, both in terms of stats and positions:
    - See sheet: https://docs.google.com/spreadsheets/d/1pocrlClqrNoDNBAIF_hZMyurMxYe5K649e1iLXPkYhI/edit?gid=1453899946#gid=1453899946

Here's an example of the merged INFO field from the sheet above:

`DB=.;ECNT=11.0;FS=0.0;HCNT=6.0;MAX_ED=80.0;MIN_ED=0.5;NLOD=175.255;NLODF=39.465;PV=0.40159999999999996;PV2=0.35085;RPA=.;RU=.;SOR=0.911;STR=.;TLOD=7.615;TNSCOPE_MNV_FILTERS=triallelic_site|in_normal,in_normal;TNSCOPE_MNV_NORMAL_ADs=494|10,726|10;TNSCOPE_MNV_NORMAL_AFs=0.0142857,0.013587;TNSCOPE_MNV_TUMOR_ADs=590|10,900|10;TNSCOPE_MNV_TUMOR_AFs=0.0115875,0.010989;TNSCOPE_MNV_VARS=2_158637135_GAAA_G,2_158637139_A_G
`
The fields added here are: 

- TNSCOPE_MNV_VARS=2_158637135_GAAA_G,2_158637139_A_G
- TNSCOPE_MNV_TUMOR_AFs=0.0115875,0.010989
- TNSCOPE_MNV_NORMAL_AFs=0.0142857,0.013587
- TNSCOPE_MNV_TUMOR_ADs=590|10,900|10
- TNSCOPE_MNV_NORMAL_ADs=494|10,726|10
- TNSCOPE_MNV_FILTERS=triallelic_site|in_normal,in_normal

### Pipeline Integrity Tests

<!-- Include tests to verify the integrity of the different Balsamic workflows. -->

- **Report deliver (generation of the `.hk` file)**
    - [x] N/A
    - [ ] Verified
- **TGA T/O Workflow**
    - [ ] N/A
    - [ ] Verified
- **TGA T/N Workflow**
    - [ ] N/A
    - [ ] Verified
- **UMI T/O Workflow**
    - [ ] N/A
    - [ ] Verified
- **UMI T/N Workflow**
    - [ ] N/A
    - [ ] Verified
- **WGS T/O Workflow**
    - [x] N/A
    - [ ] Verified
- **WGS T/N Workflow**
    - [ ] N/A
    - [x] Verified
- **QC Workflow**
    - [x] N/A
    - [ ] Verified
- **PON Workflow**
    - [x] N/A
    - [ ] Verified

## Clinical Genomics Stockholm

<!-- Do not reveal clinical data, and if applicable, place it within the internal Google Drive directory. -->

### Documentation

<!-- Link related issues or PRs for necessary changes. -->

- **Atlas documentation**
    - [x] N/A
    - [ ] Updated: [Link]
- **Web portal for Clinical Genomics**
    - [x] N/A
    - [ ] Updated: [Link]

### Panel of Normal specific criteria

<!-- If the PR includes a new PoN adhere to the criteria below. -->

- [ ] The PR includes the addition of a new Panel of Normals
- [ ] The samples have been verified to adhere to the sample selection criteria on [Atlas PoN creation instructions for Balsamic](https://atlas.scilifelab.se/infrastructure/BALSAMIC/balsamic_pon/#procedure)

### User Changes

<!-- Please provide justification if you select N/A and the changes affect the output or results. -->

- [x] N/A
- [x] This PR affects the output files or results.
    - [x] User feedback is considered unnecessary because: Users discovered the issue initially, that there were 2 representations of the same variant, then it was decided by bioinformatician in discussion with colleagues that MNV is the best way to represent variants for VEP to create the most accurate protein effect prediction .
    - [ ] Affected users have been included in the development process and given a chance to provide feedback.

### Infrastructure Changes

<!-- Link related issues or PRs for necessary changes. -->
<!-- Make sure that the linked PRs include relevant tests. -->

- **Stored files in Housekeeper**
    - [x] N/A
    - [ ] Updated: [Link]
- **CG (CLI and delivered/uploaded files)**
    - [x] N/A
    - [ ] Updated: [Link]
- **Servers (configuration files on Hasta)**
    - [x] N/A
    - [ ] Updated: [Link]
- **Scout interface**
    - [x] N/A
    - [ ] Updated: [Link]

### Validation criteria

<!-- Create validation criteria for this PR or mark as unnecessary. -->

Validation criteria to be added to validation report PR: https://github.com/Clinical-Genomics/validations/pull/285

**Version specific criteria**

In VCF of any TGA case: SNV.somatic.[case-id].tnscope.research.normalised.vcf.gz the following criteria are met:
- [ ] Created MNVs can be found
- [ ] AF,DP,FILTER,CHROM_POS_ALT_REF from the constituent SNV / InDels used to construct the MNV can be found in the MNV info-field
- [ ] Merged MNVs with filter MNV_CONFLICTING_FILTERS exist, and constituent variants do not all have matched normal filters set (in_normal/germline_risk) 
- [ ] MNV exists with filters: in_normal,germline_risk
- [ ] Merged TNscope MNV can be merged with VarDict MNV
- [ ] Merged TNscope MNV can be seen in Scout


> [!IMPORTANT]
> One of the below checkboxes for validation need to be checked 

- [x] Added version specific validation criteria to validation report
- [ ] Changes validated in standard sections: [validation-section]
- [ ] Validation criteria not necessary

## Checklist

> [!IMPORTANT]  
> Ensure that all checkboxes below are ticked before merging.

### For Developers

- **PR Description**
    - [x] Provided a comprehensive description of the PR.
    - [x] Linked relevant user stories or issues to the PR.
- **Documentation**
    - [x] Verified and updated documentation if necessary.
- **Validation criteria**
    - [x] Completed the validation criteria section of the template.
- **Tests**
    - [x] Described and tested the functionality addressed in the PR.
    - [x] Ensured integration of the new code with existing workflows.
    - [ ] Confirmed that meaningful unit tests were added for the changes introduced.
    - [x] Checked that the PR has successfully passed all relevant code smells and coverage checks.
- **Review**
    - [ ] Addressed and resolved all the feedback provided during the code review process.
    - [ ] Obtained final approval from designated reviewers.

### For Reviewers

- **Code**
    - [ ] Code implements the intended features or fixes the reported issue.
    - [ ] Code follows the project's coding standards and style guide.
- **Documentation**
    - [ ] Pipeline changes are well-documented in the CHANGELOG and relevant documentation.
- **Validation criteria**
    - [ ] The author has completed the validation criteria section of the template
- **Tests**
    - [ ] The author provided a description of their manual testing, including consideration of edge cases and boundary
      conditions where applicable, with satisfactory results.
- **Review**
    - [ ] Confirmed that the developer has addressed all the comments during the code review.

<!-- Add any other relevant information or specific checks necessary for your PR. -->

khurrammaqbool commented:
Here is the summary of my comments as we discussed.
1. It will be nice to see the stats that shows the effect of merging the variants into MNVs. This method effects TGA, TO and TN cases so some results from at least one represented case with at leas following basic stats:
     i . Number of SNVs and number of indels before merging
    ii. Number of SNVs and number of indels after merging
   iii. Number of merged variants
   iv. The ratios of the effected variants compared to the total.
   v. The Numbers and the ratio of clinically relevant variants effected by the method.

2.   It is important for the interpretation that the filter, AF, DP along with any additional information relevant for the interpretation, e.g. QUAL, should be retained for each variant merged as MNV. I suggest the following format in the INFO field of the merged VCF for this purpose: 
```
NEW_INFO_TAG:INFO_VAR1|ADDITIONAL_INFO_VAR1,INFO_VAR2|ADDITIONAL_INFO_VAR2; 
```
To make the VCF format compatible, the original INFO tags should be preserved and a new tag may be added that contains the comma separated information containing the same order as of the merged variants and has the same size as the number of merged variants to be parsed as list.
3. Scout may be able to show the additional information from the new VCF tag of the merged variants as hover or in the variant view page.
4. The PR separates the current  raw VCF and generates and moves the method to an intermediate VCF file that is not delivered, this should be mentioned somewhere in the documentation.
6. It is nice that the documentation is updated for the newly added VCF tags, but it may also be relevant to describe all the old and the new VCF tags to avoid the confusion and to be able to identify the origin of the VCF tags by TNScope or the method in the PR.

khurrammaqbool commented on BALSAMIC/assets/scripts/merge_mnp.py:
```diff
@@ -0,0 +1,430 @@
+Requirements:
+This script requires access to the vcflib library contained in the Sentieon
+software package located in $SENTIEON_INSTALL_DIR/lib/python/sentieon.
+"""
```
Please add a  summary of the changes compared to the original script from Sentieon.

mathiasbio replied:
Done!

khurrammaqbool commented on BALSAMIC/assets/scripts/merge_mnp.py:
```diff
@@ -0,0 +1,430 @@
+            "Description": '"SNV Merged with neighboring variants"',
+            "ID": "MERGED",
+        },
+        "MNV_CONFLICTING_FILTERS": {
```
This, and the subsequent TAGs, is relevant the the second point the in comment.

mathiasbio replied:
Added AF, DP, and variant chrom, pos, alt and ref for all constituent variants which were used to built the MNV

khurrammaqbool reviewed (commented):
👍 🥇  This is nice work. As I mentioned in the discussion, we should keep the merged file non-redundant i.e. all or most of the information from the variants to be merged should be retained in the merged variant and remove the additional lines representing the original variants that are merged. This will also eliminate the need for keeping the intermediate files and help with the interpretation as well.

khurrammaqbool commented on BALSAMIC/assets/scripts/merge_mnp.py:
```diff
@@ -0,0 +1,537 @@
+    return v2.pos - v1.pos - len(v1.ref) + 1
+
+
+def merge(
```
[Please consider the suggestion from sonarcloud](https://sonarcloud.io/project/issues?pullRequest=1524&open=AZSI5rlT783WF5Ea1hRG&id=Clinical-Genomics_BALSAMIC). In addition to the cognitive complexity, iterating over the list multiple times will also greatly impact the performance. Could you add some benchmarks for the scripts for us to know how much time and resources does it take for e.g. WES panel to be sure that it will not cause any problem in production when run on large panels.

mathiasbio replied:
I saw the sonarcloud issues, and I tried to fix it and ended up spending a whole day almost but somehow the script stopped working. At the moment I would like to treat this script as a mostly stand-alone-tool that we have taken from Sentieon, and where I have made adjustments where necessary but otherwise let it remain as it is. I feel at the moment with all of the pressing issues in balsamic and the very reduced manpower we have that we could allow for some cognitive complexity in this script. I'm afraid that the release will be quite delayed if I continue spending time on refactoring. 

But I will give a report on the time and resource for a WES case!

mathiasbio replied:
I added this to the PR header:

Regarding benchmarking the speed of the merge script:
- For most standard TGA cases that I checked the entire rule which includes multiple post-processing of TNscope takes 7-8 seconds
- For WES it took 22 seconds with a maximum memory peak of 50MB 

In summary it's quick and not memory-demanding

khurrammaqbool commented on BALSAMIC/assets/scripts/merge_mnp.py:
```diff
@@ -0,0 +1,537 @@
+    return vlist
+
+
+def process(
```
[Please consider this second suggestion from sonarcloud as well](https://sonarcloud.io/project/issues?pullRequest=1524&open=AZSJBL41AOgkRPOmWLTn&id=Clinical-Genomics_BALSAMIC). In addition to the cognitive complexity, iterating over the list multiple times on top of the previous function will have an additional consequences, greatly impact the performance. Could you add some benchmarks for the scripts for us to know how much time and resources does it take for e.g. WES panel to be sure that it will not cause any problem in production when run on large panels. It will be nice if some of the complexity could be reduced both from cognitive complexity and, more importantly, from performance point of view.

mathiasbio replied:
Same comment here as I wrote in the other comment 🙏

mathiasbio commented:
Thanks for the review @khurrammaqbool ! I think I have addressed all your comments and feedback : ) 

Regarding the failing docker containers I will test them all after the PR has been approved, just in case I need to make further changes to the code which will force me to restart the dockerbuilds.

khurrammaqbool commented on BALSAMIC/assets/scripts/merge_mnp.py:
```diff
@@ -0,0 +1,537 @@
+                v.info[k] = None
+            else:
+                try:
+                    v.info[k] = sum(vk) / len_vv
```
Please add some limit to the floating points. This is unnecessarily  long `PV=0.0005499999999999999;PV2=0.0005499999999999999`

mathiasbio replied:
Good point 😅 I added some round() to the script

khurrammaqbool commented on BALSAMIC/assets/scripts/merge_mnp.py:
```diff
@@ -0,0 +1,537 @@
+        _ = vcf.format(v)
+        for vi in vv:
+            _ = vcf.format(vi)
+
```
I think SOR should be also added as list here.

mathiasbio replied:
It could be nice to have. But we are doing SOR filtering in the quality filtering step before the merging of variants to MNVs, so if there was any constituent variant with a poor SOR they would be filtered out prior. At the moment I would really like to avoid needing to re-run more tests. I think it could be better to proceed with the info that has been added, and if we find that there is a need to add more info by request from the customers, then we can add those specifically later (at the moment none of these added INFO-fields are parsed in Scout)

khurrammaqbool reviewed (commented):
👍 some additional minor comments.

mathiasbio commented:
Only somalier container failed, I will open a PR to fix it!
